File: java/src/main/java/cn/az/code/year2021/june/MatchsticksToSquare.java
package cn.az.code.year2021.june;

import java.util.Arrays;

/**
 * @author az
 * @since 2021-06-15
 */
public class MatchsticksToSquare {

    public boolean makesquare(int[] matchsticks) {
        if (matchsticks == null || matchsticks.length < 4)
            return false;
        int sum = Arrays.stream(matchsticks).sum();
        if (sum % 4 != 0)
            return false;

        Arrays.sort(matchsticks);
        reverse(matchsticks);

        return dfs(matchsticks, new int[4], 0, sum / 4);
    }

    private boolean dfs(int[] matchsticks, int[] sums, int index, int target) {
        if (index == matchsticks.length) {
            if (sums[0] == target && sums[1] == target && sums[2] == target) {
                return true;
            }
            return false;
        }

        for (int i = 0; i < 4; i++) {
            if (sums[i] + matchsticks[index] > target)
                continue;
            sums[i] += matchsticks[index];
            if (dfs(matchsticks, sums, index + 1, target))
                return true;
            sums[i] -= matchsticks[index];
        }

        return false;
    }

    private void reverse(int[] matchsticks) {
        int i = 0, j = matchsticks.length - 1;
        while (i < j) {
            int temp = matchsticks[i];
            matchsticks[i] = matchsticks[j];
            matchsticks[j] = temp;
            i++;
            j--;
        }
    }

}

File: archive/go/2021/june/matchsticks-to-square.go
package june

import "sort"

func makesquare(matchsticks []int) bool {
	if len(matchsticks) < 4 {
		return false
	}
	sum := 0
	for i := range matchsticks {
		sum += i
	}
	if sum%4 != 0 {
		return false
	}
	sort.Sort(sort.Reverse(sort.IntSlice(matchsticks)))
	sli := make([]int, 4)

	return dfs(sli, matchsticks, 0, sum/4)
}

func dfs(sli []int, nums []int, index int, sum int) bool {
	if index == len(nums) {
		return sli[0] == sli[1] && sli[1] == sli[2] && sli[2] == sli[3]
	}
	for i := 0; i < 4; i++ {
		if (sli[i] + nums[index]) > sum {
			continue
		}
		j := i
		for j > 0 {
			j--
			if sli[i] == sli[j] {
				break
			}
		}
		if j != -1 {
			continue
		}
		sli[i] += nums[index]
		if dfs(sli, nums, index+1, sum) {
			return true
		}
		sli[i] -= nums[index]
	}
	return false
}
